### DISCUSSION TITLE
Terraform/OpenTofu and Ansible for Kanidm?

### DISCUSSION THREAD

**Discussion by offby0point5** [kanidm/kanidm#3155] [Polls]
I love and use Kanidm for my personal deployment. But I want to automate some tasks like the creation of OAuth2 endpoints and service accounts and the retrieval of their basic secrets/api tokens. As far as I could see, there were no signs of such automation tools yet.

So I thought I would ask if you need this too and maybe even would contribute these modules.

**Comment by Firstyear** [comment]
I think this would be great to have :)

**Reply by offby0point5** [reply]
Can you imagine those to be part of the Kanidm organization but maintained by others using Kanidm, so they may be a bit like semi-official?

**Reply by Firstyear** [reply]
Yes, I can, I don't have an issue with that. @yaleman ? 

Maybe the best place is to start with the opentofu / ansible repos in your personal repo, and then if you want later we can transfer them into Kanidm org ?

**Reply by yaleman** [reply]
Sounds sensible to me

**Reply by Firstyear** [reply]
Okay, sounds good. @offby0point5 If you start the projects elsewhere first, we're happy to adopt them into the Kanidm org later once they get a bit more established. Is that okay?

**Reply by offby0point5** [reply]
Okay, I will try my best. Is it okay to link the repos here?

**Reply by Firstyear** [reply]
Yeah, of course you can!

**Comment by dkowis** [comment]
I love that there's OpenTofu(Terraform) scripts for Keycloak, and I'll miss them coming to kanidm, but I think switching away from a complex LDAP setup might be worth it.

**Reply by offby0point5** [reply]
The LDAP setup is why I chose Kanidm too.
I will start with an OpenTofu provider, but it might take a while as I have to learn how to build a provider first.

**Reply by Firstyear** [reply]
We can't wait to see what you create, it's really appreciated.

**Reply by almereyda** [reply]
I've made good experiences with using code generation from an OpenAPI schema to scaffold a Terraform/OpenTofu provider.

- [REST Interface - Kanidm Administration](https://kanidm.github.io/kanidm/stable/developers/designs/rest_interface.html)
- [Code generation | Terraform | HashiCorp Developer](https://developer.hashicorp.com/terraform/plugin/code-generation)
- [OpenAPI provider spec generator | Terraform | HashiCorp Developer](https://developer.hashicorp.com/terraform/plugin/code-generation/openapi-generator)

**Reply by KittyKatt** [reply]
Did anyone make any progress on this yet? I've dried up a lot of the things I wanted to work on with my first tf provider and now I'm itching to do another one.

**Reply by yaleman** [reply]
There's Rust crates for generating tofu providers now that seem... workable, I'd happily poke around at helping with it.. I (wo|ca)n't code in golang (the primary terraform/tofu provider language)  though 😄

**Reply by dkowis** [reply]
I hear that, golang makes me break out in expletives, just like python :(

**Comment by offby0point5** [comment]
I finally started creating an [OpenTofu provider](https://github.com/slop-incubator/terraform-provider-kanidm). Admittedly it's initially generated by LLM (just like other existing ones) but I plan to not use LLMs further on and learn by 'deslop' how everything works.
In addition, I started a [Go client library](https://github.com/slop-incubator/go-kanidm) (also by LLM, no-AI further on) which looks like to be the only one right now, as [other Kanidm projects in Go](https://github.com/search?q=kanidm+language%3AGo&type=repositories&l=Go) seem to have it baked in.

If you'd like to help and endure some pain, it's appreciated :)

**Reply by KittyKatt** [reply]
Unrelated to Terraform/OpenTofu/Ansible, I'd be interested to see where that Go client library goes just for the [kanidm-operator](https://github.com/KittyKatt/kanidm-operator) I've been creating.
